### PR TITLE
Adding extended RTMP support

### PR DESCRIPTION
**Pull request by rogerhu**
The Extended RTMP standard (https://github.com/veovera/enhanced-rtmp ) relies on the unused highest order bit of the video tag frame byte. If this bit is set to 1, then another 4 bytes are read from the stream, which will denote the codec type (e.g. AV1, HVEC, VP9). In addition, the existing lower half byte is then used to provide signaling information (e.g. packet sequence start, coded frames, MPEG2 TS frames, etc.)

Inspired by https://github.com/ossrs/srs/pull/3495/files#diff-b381246f7b6c2af916111db040dc6a1d52a5288f4c0d2b783f5a90b13ac0d5b8

You can also create your own FLV with H.265 encoding with FFMPEG 6.1 or above:

```
pushd /tmp
wget https://sample-videos.com/video123/mp4/720/big_buck_bunny_720p_1mb.mp4
ffmpeg -i /tmp/big_buck_bunny_720p_1mb.mp4 -c:a aac -vcodec libx265 test.flv    
```

I'm also able to test by creating an FLV file from using this Android library that incorporates these changes:

https://github.com/ThibaultBee/StreamPack/commit/3b4fe5b414f1b60f31cd2636ba45b1040445d1b4

### CODE REVIEW

**Comment by noirotm** [comment]
Hi Roger!
Thanks for this PR, this is pretty interesting to see FLV and RTMP still alive and evolving in 2023.
I see that this is marked as WIP. What's missing before it can be merged?

On my side, Since I wasn't aware of these latest evolutions, I need to take a look at the spec to understand it, then I'll try to review the proposed changes to FLVMeta.

Also, would you please be able to share sample video files demonstrating these changes?

Regards,
Marc

**Comment by rogerhu** [comment]
Hi Marc,

I added a few unit tests to try to show the changes. Take a look!

I've also included an FLV file as part of the PR for review purposes. I can remove it before it's merged!

**Comment by noirotm** [comment]
FYI, I just pushed a commit to fix the audio/video typo you also fixed here.
Reason is I just wanted to make sure the "build" github action still worked as expected, as the run cache was empty.
Sorry this causes you to have to resolve the conflict 😅.

Please note that on Windows, which is a supported platform for FLVMeta, the current build script action does not support the Check-based tests yet.
However, on my Windows dev machine, I usually check out Check separately so I don't rely on find_package.
The bottom line is that I'm afraid the newly added `#include <unistd.h>` and function usage will cause the build for the tests to fail on Windows.

One solution could be to add a CMake check for the unistd.h header and conditionally compile the tests you added so the old ones can still be run.

**Comment by rogerhu** [comment]
> One solution could be to add a CMake check for the unistd.h header and conditionally compile the tests you added so the old ones can still be run.

I just stopped using unistd.h and moved everything to use fwrite()!

**Comment by rogerhu** [comment]
Official ffmpeg changes are here: https://github.com/FFmpeg/FFmpeg/commit/aa83654d940e7599ee641fc5d0fcd10e7decf277

FFMPeg 6.1 includes the enhanced RTMP spec (https://9to5linux.com/ffmpeg-6-1-heaviside-released-with-vaapi-av1-encoder-hw-vulkan-decoding)

Running 6.1 version, we can also create an FLV with this new standard:

```
ffmpeg -i /tmp/sample_1280x720_surfing_with_audio.mp4 -c:a aac -vcodec libx265 test.flv    
```

**Comment by rogerhu** [comment]
Thanks to this tool, I found a minor bug in the ffmpeg implementation - submitting a patch to [address](https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/flvenc.c#L542-L547) - should show up on https://ffmpeg.org/pipermail/ffmpeg-devel/2023-November/thread.html:

Code changes proposed:
```
--- a/libavformat/flvenc.c
+++ b/libavformat/flvenc.c
@@ -540,10 +540,10 @@ static void
flv_write_codec_header(AVFormatContext* s, AVCodecParameters* par, i
             avio_write(pb, par->extradata, par->extradata_size);
         } else {
             if (par->codec_id == AV_CODEC_ID_HEVC) {
-                avio_w8(pb, FLV_IS_EX_HEADER |
PacketTypeSequenceStart | FLV_FRAME_KEY); // ExVideoTagHeader mode
with PacketTypeSequenceStart
+                avio_w8(pb, FLV_IS_EX_HEADER |
(PacketTypeSequenceStart << FLV_VIDEO_FRAMETYPE_OFFSET)); //
ExVideoTagHeader mode with PacketTypeSequenceStart
                 avio_write(pb, "hvc1", 4);
```

Running `build/src/flvmeta -r -F /tmp/test.flv`:

Before:
```
--- Tag #2 at 0x190 (400) ---

* Video codec: HEVC
* Video frame type: coded frames
```

After:

```
--- Tag #2 at 0x190 (400) ---

* Video codec: HEVC
* Video frame type: sequence start
```

**Comment by rogerhu** [comment]
> Thanks to this tool, I found a minor bug in the ffmpeg implementation - submitting a patch to [address](https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/flvenc.c#L542-L547) - should show up on https://ffmpeg.org/pipermail/ffmpeg-devel/2023-November/thread.html:

Disregard. I found a bug after re-reading the spec and realized that the packet type is the lower 4 bits. Pushed the fix.

**Review by noirotm — CHANGES_REQUESTED** [review]
There are a few things that should be changed, as noted in the individual comments.

I propose to keep the video tag a single type that abstracts the difference between legacy and extended.

Also, the FOURCC constants and the related could should probably be `uint32_be` to make them easier to parse and store.

In addition, I believe there are things that are not implemented in this PR that can be added later (unless I missed them):
- it doesn't look like the metadata updating code takes into account the FOURCC codec yet. It is part of the updated spec to write the value in the `videocodecid` property.
- some checks probably need to be added to take the new headers into account.
- video size computation needs to be implemented for the newly added codecs

Apart from these small issues, great job, thank you for the work put into this PR 👍 FLVMeta isn't the easiest codebase to get into, and it's been around for quite a while.

File: src/flv.h [hunk 1]
```diff
@@ -132,10 +132,34 @@ typedef byte flv_audio_tag;
...
+typedef struct __flv_video_tag_ext {
+    byte fourcc[4];
+} flv_video_tag_ext;
+
+typedef byte flv_video_ext_codec;
+
+#define FLV_VIDEO_FOURCC_AV1        { 'a', 'v', '0', '1' }
+#define FLV_VIDEO_FOURCC_VP09       { 'v', 'p', '0', '9' }
+#define FLV_VIDEO_FOURCC_HEVC       { 'h', 'v', 'c', '1' }
+#define FLV_VIDEO_FOURCC_SIZE 4
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.h:164]: Thanks for the suggestion!

File: src/flv.c [hunk 2]
```diff
@@ -182,7 +182,7 @@ int flv_read_audio_tag(flv_stream * stream, flv_audio_tag * tag) {
     return FLV_OK;
 }
 
-int flv_read_video_tag(flv_stream * stream, flv_video_tag * tag) {
+int _flv_read_video_tag(flv_stream * stream, flv_video_tag * tag) {
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.c]: Consolidated back!

File: src/flv.h [hunk 3]
```diff
@@ -132,10 +132,34 @@ typedef byte flv_audio_tag;
...
+
+/* See https://github.com/veovera/enhanced-rtmp */
+typedef struct __flv_video_tag_ext {
+    byte fourcc[4];
+} flv_video_tag_ext;
+
+typedef byte flv_video_ext_codec;
+
+#define FLV_VIDEO_FOURCC_AV1        { 'a', 'v', '0', '1' }
+#define FLV_VIDEO_FOURCC_VP09       { 'v', 'p', '0', '9' }
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.h]: Changed! Thanks for the catch.

File: src/dump.c [hunk 4]
```diff
@@ -27,6 +27,42 @@
 
 #include <string.h>
 
+int parse_ext_codec(flv_video_tag tag, flv_video_tag_ext tag_ext, flv_video_ext_codec* ext_codec) {
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/dump.c]: I got rid of this function now!

File: src/flv.c [hunk 5]
```diff
@@ -216,6 +216,46 @@ int flv_read_video_tag(flv_stream * stream, flv_video_tag * tag) {
     return FLV_OK;
 }
 
+int flv_read_fourcc_codec(flv_stream * stream, flv_video_tag_ext *ext) {
+    char FLV_VIDEO_TAG_FOURCC_AV1[] = FLV_VIDEO_FOURCC_AV1;
+    char FLV_VIDEO_TAG_FOURCC_VP9[] = FLV_VIDEO_FOURCC_VP09;
+    char FLV_VIDEO_TAG_FOURCC_HEVC[] = FLV_VIDEO_FOURCC_HEVC;
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.c]: Unused - deleted.

File: src/dump_xml.c [hunk 6]
```diff
@@ -175,41 +175,45 @@ static int xml_on_tag(flv_tag * tag, flv_parser * parser) {
     return OK;
 }
 
-static int xml_on_video_tag(flv_tag * tag, flv_video_tag vt, flv_parser * parser) {
-    printf("    <videoData codecID=\"%s\"", dump_string_get_video_codec(vt));
+static int xml_on_video_tag(flv_tag * tag, flv_video_tag vt, flv_video_tag_ext vt_ext, flv_parser * parser) {
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/dump_xml.c]: Updated! I think?

File: src/check.c [hunk 7]
```diff
@@ -542,26 +546,41 @@ int check_flv_file(const flvmeta_opts * opts) {
...
+                if(ext_header == 1) {
+                    flv_video_ext_codec ext_codec;
+                    int retval = parse_ext_codec(vt, vt_ext, &ext_codec);
+                    if (retval == -1) {
+                        sprintf(message, "unknown extended video codec %c%c%c%c",
+                        vt_ext.fourcc[0],
+                        vt_ext.fourcc[1],
+                        vt_ext.fourcc[2],
+                        vt_ext.fourcc[3]);
+                        print_error(ERROR_VIDEO_CODEC_UNKNOWN, offset, message);
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/check.c]: Changed to ERROR_EXTENDED_VIDEO_CODEC_UNKNOWN

**Comment by rogerhu** [comment]
> * it doesn't look like the metadata updating code takes into account the FOURCC codec yet. It is part of the updated spec to write the value in the `videocodecid` property.

Ah yes! Good catch.

> * some checks probably need to be added to take the new headers into account.

Yeah and there some additional warnings I think in general.

> * video size computation needs to be implemented for the newly added codecs

Good call-out! 
 
> Apart from these small issues, great job, thank you for the work put into this PR 👍 FLVMeta isn't the easiest codebase to get into, and it's been around for quite a while.

Thank you for the helpful feedback!

**Comment by noirotm** [comment]
Sorry for the lack of replies this week, life got in the way.
I'll review the latest changes this weekend to hopefully merge this PR 🤞

File: src/dump.c [hunk 8]
```diff
@@ -72,6 +87,23 @@ const char * dump_string_get_video_frame_type(flv_video_tag tag) {
...
 
+const char * dump_string_get_ext_packet_type(flv_video_tag tag) {
+    if (flv_video_tag_is_ext_header(tag.video_tag)) {
+       int packet_type = flv_video_tag_ext_packet_type(tag.video_tag);
+
+       switch (packet_type) {
+           case FLV_VIDEO_TAG_PACKET_TYPE_SEQUENCE_START: return "sequence start";
+           case FLV_VIDEO_TAG_PACKET_TYPE_CODED_FRAMES: return "coded frames";
+           case FLV_VIDEO_TAG_PACKET_TYPE_SEQUENCE_END: return "sequence end";
+           case FLV_VIDEO_TAG_PACKET_TYPE_CODED_FRAMES_X: return "coded frames";
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/dump.c]: thanks, fixed

File: src/flv.c [hunk 9]
```diff
@@ -182,38 +182,58 @@ int flv_read_audio_tag(flv_stream * stream, flv_audio_tag * tag) {
...
+        int fourcc_retval = flv_read_fourcc_codec(stream, tag);
+
+        if (fourcc_retval == FLV_ERROR_EOF) {
+            flv_close(stream);
+            return FLV_ERROR_EOF;
         }
     }
 
-    return FLV_OK;
+  return FLV_OK;
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.c]: yep fixed!

File: src/flv.c [hunk 10]
```diff
@@ -486,8 +507,9 @@ int flv_parse(const char * file, flv_parser * parser) {
                 flv_close(parser->stream);
                 return retval;
             }
+
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.c]: fixed i think

File: src/check.c [hunk 11]
```diff
@@ -542,25 +544,33 @@ int check_flv_file(const flvmeta_opts * opts) {
...
-                    && video_codec != FLV_VIDEO_TAG_CODEC_ON2_VP6
-                    && video_codec != FLV_VIDEO_TAG_CODEC_ON2_VP6_ALPHA
-                    && video_codec != FLV_VIDEO_TAG_CODEC_SCREEN_VIDEO_V2
-                    && video_codec != FLV_VIDEO_TAG_CODEC_AVC
-                ) {
-                    sprintf(message, "unknown video codec id %u", video_codec);
-                    print_error(ERROR_VIDEO_CODEC_UNKNOWN, offset + 11, message);
-                }
+                /* check video codec first by checking if ext_header is set */
+                if(flv_video_tag_is_ext_header(vt.video_tag)) {
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/check.c]: fixed

File: src/flv.c [hunk 12]
```diff
@@ -486,8 +507,9 @@ int flv_parse(const char * file, flv_parser * parser) {
...
             if (retval != FLV_ERROR_EMPTY_TAG && parser->on_video_tag != NULL) {
-                retval = parser->on_video_tag(&tag, vt, parser);
+	        retval = parser->on_video_tag(&tag, vt, parser);
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.c]: fixed, and made spaces the default now in my editor

File: src/check.c [hunk 13]
```diff
@@ -517,12 +518,13 @@ int check_flv_file(const flvmeta_opts * opts) {
                 }
 
                 /* check whether the format varies between tags */
-                if (have_prev_video_tag && flv_video_tag_codec_id(prev_video_tag) != flv_video_tag_codec_id(vt)) {
+                if (have_prev_video_tag && flv_video_tag_codec_id(prev_video_tag.video_tag) != flv_video_tag_codec_id(vt.video_tag)) {
                     print_warning(WARNING_VIDEO_FORMAT_CHANGED, offset + 11, "video format changed since last tag");
                 }
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/check.c]: made flv_video_tag_codec_id work for both!

File: src/flv.c [hunk 14]
```diff
@@ -331,6 +351,7 @@ void flv_reset(flv_stream * stream) {
     }
 }
 
+
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.c]: i went through and try to find all extra lines added

File: src/dump.c [hunk 15]
```diff
@@ -37,29 +38,43 @@ const char * dump_string_get_tag_type(flv_tag * tag) {
...
+    if (flv_video_tag_is_ext_header(tag.video_tag))
+    {
+        if (tag.fourcc == FLV_VIDEO_FOURCC_HEVC)
+            return "HEVC";
+        else if (tag.fourcc == FLV_VIDEO_FOURCC_AV1)
+            return "AV1";
+        else if (tag.fourcc == FLV_VIDEO_TAG_CODEC_VP9)
+            return "VP9";
+        else
+            return "Unknown";
```
[the remark this answers is not in the document]
Comment (noirotm) [reply src/dump.c]: Oh I see, I hadn't realized there were both
```C
/* Extended RTMP codecs */
#define FLV_VIDEO_TAG_CODEC_AV1             99
#define FLV_VIDEO_TAG_CODEC_VP9             100
#define FLV_VIDEO_TAG_CODEC_HEVC            101
```
and
 ```C
#define FLV_VIDEO_FOURCC_AV1        FOURCC('a', 'v', '0', '1')
#define FLV_VIDEO_FOURCC_VP9        FOURCC('v', 'p', '0', '9')
#define FLV_VIDEO_FOURCC_HEVC       FOURCC('h', 'v', 'c', '1')
```

Given that the latter are now defined as integers, you probably want to use them instead of the former, since their value seems to be arbitrary.

In the end, I think the codec list could be:

```C
/* video tag */
#define FLV_VIDEO_TAG_CODEC_JPEG            1
#define FLV_VIDEO_TAG_CODEC_SORENSEN_H263   2
#define FLV_VIDEO_TAG_CODEC_SCREEN_VIDEO    3
#define FLV_VIDEO_TAG_CODEC_ON2_VP6         4
#define FLV_VIDEO_TAG_CODEC_ON2_VP6_ALPHA   5
#define FLV_VIDEO_TAG_CODEC_SCREEN_VIDEO_V2 6
#define FLV_VIDEO_TAG_CODEC_AVC             7

/* Extended RTMP codecs */
#define FLV_VIDEO_FOURCC_AV1        FOURCC('a', 'v', '0', '1')
#define FLV_VIDEO_FOURCC_VP9        FOURCC('v', 'p', '0', '9')
#define FLV_VIDEO_FOURCC_HEVC       FOURCC('h', 'v', 'c', '1')
```

so there's no confusion.

[the remark this answers is not in the document]
Comment (rogerhu) [reply src/dump.c]: Ah gotcha, good catch. I changed.

File: src/flv.h [hunk 16]
```diff
@@ -132,10 +132,34 @@ typedef byte flv_audio_tag;
...
-#define flv_video_tag_codec_id(tag)     (((tag) & 0x0F) >> 0)
-#define flv_video_tag_frame_type(tag)   (((tag) & 0xF0) >> 4)
+
+#define flv_video_tag_codec_id(tag)          (((tag) & 0x0F) >> 0)
+#define flv_video_tag_frame_type(tag)        (((tag) & 0x70) >> 4)
+
+/* See https://github.com/veovera/enhanced-rtmp */
+typedef struct __flv_video_tag_ext {
+    byte fourcc[4];
+} flv_video_tag_ext;
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply src/flv.h]: Done!

File: schemas/flv.xsd [hunk 17]
```diff
@@ -108,6 +108,7 @@
         </xs:choice>
         <xs:attribute name="frameType" use="required" type="tFrameType"/>
         <xs:attribute name="codecID" use="required" type="tVideoCodec"/>
+        <xs:attribute name="packetType" type="tPacketType"/>
```
[the remark this answers is not in the document]
Comment (rogerhu) [reply schemas/flv.xsd]: Thanks fixed.

**Comment by rogerhu** [comment]
> > * it doesn't look like the metadata updating code takes into account the FOURCC codec yet. It is part of the updated spec to write the value in the `videocodecid` property.

It looks like I just needed to update the info.h header to 32-bits to make things work! Pushed the changes.

```
--- a/src/info.h
+++ b/src/info.h
@@ -30,7 +30,7 @@ typedef struct __flv_info {
     uint8 have_audio;
     uint32 video_width;
     uint32 video_height;
-    uint8 video_codec;
+    uint32 video_codec;
     uint32 video_frames_number;
     uint8 audio_codec;
     uint8 audio_size;
```

**Comment by rogerhu** [comment]
FYI - Google has a talk about the changes here: https://www.youtube.com/watch?v=5745HMvARRs&t=797s

**Review by noirotm — COMMENTED** [review]
LGTM 👍
